### ISSUE TITLE
no-extra-parens false positive on (a&b)?c:d

### ISSUE THREAD

**Issue by andoras** [eslint/eslint#9654]
Latest NodeJS and NPM LTS 64 Bit and ESLint 4.11 on Win7 64Bit
ESLint parser "espree"

<details>
<summary>Configuration</summary>
{
	"parser": "espree",
	"parserOptions": {
		"ecmaVersion": 5,
		"sourceType": "script"
	},

	"plugins": ["angular", "filenames", "html"],
	"settings": {
		"html/xml-mode": true
	},
	"env": {
		"browser": true,
		"mocha": true,
		"node": true,
		"jasmine": true
	},
	"globals": {
		"angular": true,
		"module": true,
		"_": true,
		"$": true,
		"jasmine": true,
		"spyOn": true,
		"expect": true,
		"inject": true,
		"datajs": true,
		"SWAC": true,
		"UmcWebSSO": true,
		"uuid": true,
		"PWC": true,
		"mxGraph": true,
		"mxEventObject": true,
		"mxCell": true,
		"mxCellOverlay": true,
		"mxPoint": true,
		"mxGraphSelectionModel": true,
		"mxEvent": true,
		"mxUtils": true,
		"mxCellHighlight": true,
		"mxConstants": true,
		"mxImage": true,
		"mxRectangle": true,
		"mxRectangleShape": true,
		"mxLayoutManager": true,
		"mxClient": true,
		"mxVertexHandler": true,
		"mxImageShape": true,
		"mxEventSource": true,
		"mxMouseEvent": true,
		"mxRubberband": true,
		"mxPanningHandler": true,
		"mxGraphModel": true,
		"mxUndoManager": true,
		"mxCompactTreeLayout": true,
		"mxPopupMenuHandler": true,
		"mxDefaultPopupMenu": true,
		"mxRootChange": true,
		"mxValueChange": true,
		"mxCellAttributeChange": true,
		"mxGeometry": true,
		"mxSwimlaneManager": true,
		"mxStackLayout": true,
		"mxShape": true,
		"mxCellRenderer": true,
		"mxText": true,
		"mxCellEditor": true,
		"mxGraphHandler": true,
		"mxEdgeStyle": true,
		"mxGraphView": true,
		"mxConstraintHandler": true,
		"mxPolyline": true,
		"mxCellMarker": true,
		"mxConnectionConstraint": true,
		"mxKeyHandler": true,
		"define": true,
		"require": false,
		"factory": true,
		"Map": true
	},
	"rules": {
		"for-direction": "error",
		"no-await-in-loop": "error",
		"no-compare-neg-zero": "error",
		"no-cond-assign": "error",
		"no-console": "warn",
		"no-constant-condition": "error",
		"no-control-regex": "error",
		"no-debugger": "error",
		"no-dupe-args": "error",
		"no-dupe-keys": "error",
		"no-duplicate-case": "error",
		"no-empty-character-class": "warn",
		"no-empty": "warn",
		"no-ex-assign": "error",
		"no-extra-boolean-cast": "error",
		"no-extra-parens": [
			"error",
			"all", {
				"nestedBinaryExpressions": false
			}
		],
		"no-extra-semi": "warn",
		"no-func-assign": "error",
		"no-inner-declarations": "error",
		"no-invalid-regexp": "error",
		"no-irregular-whitespace": "error",
		"no-obj-calls": "error",
		"no-proto": "error",
		"no-prototype-builtins": "error",
		"no-regex-spaces": "error",
		"no-sparse-arrays": "error",
		"no-template-curly-in-string": "error",
		"no-unexpected-multiline": "error",
		"no-unreachable": "error",
		"no-unsafe-finally": "error",
		"no-unsafe-negation": "error",
		"use-isnan": "error",
		"valid-jsdoc": "error",
		"valid-typeof": "error",

		"accessor-pairs": "warn",
		"array-callback-return": "error",
		"callback-return": "error",
		"class-methods-use-this": "error",
		"complexity": ["error", 20],
		"consistent-return": [
			"error", {
				"treatUndefinedAsUnspecified": true
			}
		],
		"curly": ["error", "all"],
		"default-case": "error",
		"dot-location": [ "warn", "property"],
		"dot-notation": "warn",
		"eqeqeq": ["error", "always"],
		"guard-for-in": "error",
		"no-alert": "error",
		"no-caller": "error",
		"no-case-declarations": "error",
		"no-div-regex": "error",
		"no-else-return": "warn",
		"no-empty-function": "error",
		"no-empty-pattern": "error",
		"no-eq-null": "error",
		"no-eval": "error",
		"no-extend-native": "error",
		"no-extra-bind": "error",
		"no-extra-label": "error",
		"no-fallthrough": "error",
		"no-floating-decimal": "warn",
		"no-global-assign": "error",
		"no-implicit-coercion": "warn",
		"no-implicit-globals": "error",
		"no-implied-eval": "error",
		"no-invalid-this": "error",
		"no-iterator": "error",
		"no-labels": "error",
		"no-lone-blocks": "error",
		"no-loop-func": "error",
		"no-magic-numbers": ["error", {
				"ignore": [0]
			}
		],
		"no-multi-spaces": "warn",
		"no-multi-str": "error",
		"no-new-func": "error",
		"no-new-wrappers": "error",
		"no-new": "error",
		"no-octal-escape": "error",
		"no-octal": "error",
		"no-param-reassign": "error",
		"no-redeclare": "error",
		"no-return-assign": "error",
		"no-return-await": "error",
		"no-script-url": "error",
		"no-self-assign": "error",
		"no-self-compare": "error",
		"no-sequences": "error",
		"no-throw-literal": "error",
		"no-unmodified-loop-condition": "error",
		"no-unused-expressions": "error",
		"no-unused-labels": "error",
		"no-useless-call": "error",
		"no-useless-concat": "error",
		"no-useless-escape": "error",
		"no-useless-return": "error",
		"no-void": "error",
		"no-warning-comments": ["error", {
				"terms": ["todo", "fixme", "hack", "xxx", "undone", "bug", "wtf"],
				"location": "start"
			}
		],
		"no-with": "error",
		"prefer-promise-reject-errors": "warn",
		"radix": ["error", "always"],
		"require-await": "error",
		"vars-on-top": "warn",
		"wrap-iife": "warn",
		"yoda": "warn",

		"arrow-body-style": "off",
		"arrow-parens": "off",
		"arrow-spacing": "off",
		"block-scoped-var": "off",
		"constructor-super": "off",
		"generator-star-spacing": "off",
		"no-class-assign": "off",
		"no-confusing-arrow": "off",
		"no-const-assign": "off",
		"no-dupe-class-members": "off",
		"no-duplicate-imports": "off",
		"no-new-symbol": "off",
		"no-restricted-imports": "off",
		"no-this-before-super": "off",
		"no-useless-computed-key": "off",
		"no-useless-constructor": "off",
		"no-useless-rename": "off",
		"no-var": "off",
		"object-shorthand": "off",
		"prefer-arrow-callback": "off",
		"prefer-const": "off",
		"prefer-destructuring": "off",
		"prefer-numeric-literals": "off",
		"prefer-reflect": "off",
		"prefer-rest-params": "off",
		"prefer-spread": "off",
		"prefer-template": "off",
		"require-yield": "off",
		"rest-spread-spacing": "off",
		"sort-imports": "off",
		"strict": "off",
		"symbol-description": "off",
		"template-curly-spacing": "off",
		"yield-star-spacing": "off",

		"array-bracket-newline": ["warn", {
				"multiline": true,
				"minItems": 3
			}
		],
		"array-bracket-spacing": "warn",
		"array-element-newline": "warn",
		"block-spacing": "warn",
		"brace-style": "warn",
		"camelcase": "warn",
		"capitalized-comments": "off",
		"comma-dangle": "warn",
		"comma-spacing": "error",
		"comma-style": "error",
		"computed-property-spacing": "warn",
		"consistent-this": ["warn"],
		"eol-last": "warn",
		"func-call-spacing": "warn",
		"func-name-matching": "warn",
		"func-names": "warn", 
		"func-style": "warn",
		"id-blacklist": "warn",
		"id-length": "warn",
		"id-match": "warn",
		"indent": "warn",
		"jsx-quotes": ["error", "prefer-single"],
		"key-spacing": "warn",
		"keyword-spacing": "warn",
		"linebreak-style": "warn",
		"line-comment-position": "warn",
		"lines-around-comment": ["warn", {
				"allowBlockStart": true,
				"allowObjectStart": true
			}
		],
		// Deprecated since 4.0    "lines-around-directive": "warn",
		"max-depth": ["warn", 4],
		"max-len": [
			"warn", {
				"code": 160,
				"ignoreUrls": true,
				"tabWidth": 4
			}
		],
		"max-lines": "warn",
		"max-nested-callbacks": ["warn", 3],
		"max-params": ["warn", 7],
		"max-statements": [
			"warn",
			10, {
				"ignoreTopLevelFunctions": true
			}
		],
		"max-statements-per-line": [
			"warn", {
				"max": 1
			}
		],
		"lines-between-class-members": "warn",
		"multiline-comment-style": [ "warn", "starred-block"],
		"multiline-ternary": "warn",
		"new-cap": "warn",
		"new-parens": "warn",
		// Deprecated since 4.0    "newline-after-var": "warn",
		// Deprecated since 4.0    "newline-before-return": "warn",
		"newline-per-chained-call": "warn",
		"no-array-constructor": "warn",
		"no-bitwise": "error",
		"no-continue": "error",
		"function-paren-newline": "warn",
		"no-inline-comments": "warn",
		"no-lonely-if": "warn",
		"no-mixed-operators": "warn",
		"no-mixed-spaces-and-tabs": "error",
		"no-multi-assign": "warn",
		"no-multiple-empty-lines": "warn",
		"no-negated-condition": "warn",
		"no-nested-ternary": "warn",
		"no-new-object": "warn",
		"no-plusplus": ["warn", {
				"allowForLoopAfterthoughts": true
			}
		],
		"no-restricted-syntax": "warn",
		"no-tabs": "error",
		"no-ternary": "warn",
		"no-trailing-spaces": "warn",
		"no-underscore-dangle": "warn",
		"no-unneeded-ternary": "warn",
		"no-whitespace-before-property": "warn",
		"nonblock-statement-body-position": "warn",
		"object-curly-newline": ["error", {
				"minProperties": 1
			}
		],
		"object-curly-spacing": "error",
		"object-property-newline": "warn",
		"one-var-declaration-per-line": "warn",
		"one-var": "warn",
		"operator-assignment": "warn",
		"operator-linebreak": "warn",
		"padded-blocks": ["warn", "never"],
		"padding-line-between-statements":
		["warn", {
				"blankLine": "always",
				"prev": "block-like",
				"next": "*"
			}, {
				"blankLine": "always",
				"prev": "*",
				"next": "class"
			}, {
				"blankLine": "always",
				"prev": "directive",
				"next": "*"
			}, {
				"blankLine": "always",
				"prev": "export",
				"next": "*"
			}, {
				"blankLine": "always",
				"prev": "import",
				"next": "*"
			}, {
				"blankLine": "always",
				"prev": "*",
				"next": "return"
			}
		],
		"quote-props": ["error", "as-needed"],
		"quotes": ["error", "single"],
		"require-jsdoc": "warn",
		"semi-spacing": [
			"error", {
				"before": false,
				"after": true
			}
		],
		"semi": ["error", "always"],
		"semi-style": ["warn", "last"],
		"sort-keys": "off",
		"sort-vars": "off",
		"space-before-blocks": ["warn", "always"],
		"space-before-function-paren": "warn",
		"space-in-parens": "warn",
		"space-infix-ops": "warn",
		"space-unary-ops": "warn",
		"spaced-comment": "warn",
		"switch-colon-spacing": ["error", {
				"after": true,
				"before": false
			}
		],
		"template-tag-spacing": "warn",
		"unicode-bom": ["error", "never"],
		"wrap-regex": "warn",

		"global-require": "warn",
		"handle-callback-err": "warn",
		"no-buffer-constructor": "warn",
		"no-mixed-requires": "warn",
		"no-new-require": "warn",
		"no-path-concat": "warn",
		"no-process-env": "warn",
		"no-process-exit": "warn",
		"no-restricted-modules": "warn",
		"no-restricted-properties": "warn",
		"no-sync": "warn",

		"init-declarations": "off",
		"no-catch-shadow": "error",
		"no-delete-var": "error",
		"no-label-var": "error",
		"no-restricted-globals": "error",
		"no-shadow-restricted-names": "error",
		"no-shadow": "error",
		"no-undef-init": "warn",
		"no-undef": "error",
		"no-undefined": "off",
		"no-unused-vars": "error",
		"no-use-before-define": ["error", {
				"functions": true,
				"classes": true
			}
		],

		"angular/avoid-scope-typos": "off", /* JS Parsers error with this rule enabled */
		"angular/module-getter": "off", /* seems buggy atm */
		"angular/module-setter": "off", /* seems buggy atm */
		"angular/no-private-call": "error",

		"angular/angularelement": "warn",
		"angular/definedundefined": "off",
		"angular/document-service": "warn",
		"angular/foreach": "warn",
		"angular/interval-service": "warn",
		"angular/json-functions": "warn",
		"angular/log": "warn",
		"angular/no-angular-mock": "warn",
		"angular/no-jquery-angularelement": "warn",
		"angular/timeout-service": "warn",
		"angular/typecheck-array": "warn",
		"angular/typecheck-date": "warn",
		"angular/typecheck-function": "warn",
		"angular/typecheck-number": "warn",
		"angular/typecheck-object": "warn",
		"angular/typecheck-string": "warn",
		"angular/window-service": "warn",

		"angular/component-limit": "error",
		"angular/controller-as-route": "error",
		"angular/controller-as-vm": "error",
		"angular/controller-as": "error",
		"angular/deferred": "error",
		"angular/di-unused": "warn",
		"angular/directive-restrict": "warn",
		"angular/empty-controller": "warn",
		"angular/no-controller": "error",
		"angular/no-inline-template": "warn",
		"angular/no-run-logic": "warn",
		"angular/no-services": "warn",
		"angular/on-watch": "error",
		"angular/prefer-component": "warn",

		"angular/component-name": "warn",
		"angular/controller-name": "warn",
		"angular/constant-name": "warn",
		"angular/directive-name": "warn",
		"angular/factory-name": "warn",
		"angular/file-name": "warn",
		"angular/filter-name": "warn",
		"angular/module-name": "warn",
		"angular/service-name": "warn",
		"angular/provider-name": "warn",
		"angular/value-name": "warn",

		"angular/di-order": "warn",
		"angular/di": "warn",
		"angular/dumb-inject": "warn",
		"angular/function-type": ["warn", "named"],
		"angular/module-dependency-order": "warn",
		"angular/no-service-method": "warn",
		"angular/one-dependency-per-line": "warn",
		"angular/rest-service": "warn",
		"angular/watchers-execution": "warn",

		"filenames/match-regex": ["error", "^[a-z]+$", true],
		"filenames/match-exported": "warn",
		"filenames/no-index": "warn"
	}
}
</details>

I used this simple JS example:
```js
a = (b && c) ? d : e;
```

And this ESLint command
```bash
eslint --config rules test.js
```

**What should happen**
ESLint should not raise a issue for that code regarding parenthesis around '(b && c)'.

<details>
<summary>Output</summary>
<code>

1:1   error    Implicit global variable, assign as global property instead              no-implicit-globals

1:1   error    'a' is not defined                                                       no-undef

1:5   error    Gratuitous parentheses around expression                                 no-extra-parens

1:5   warning  Ternary operator used                                                    no-ternary

1:6   warning  Expected newline between test and consequent of ternary expression       multiline-ternary

1:6   error    'b' is not defined                                                       no-undef

1:11  error    'c' is not defined                                                       no-undef

1:16  warning  Expected newline between consequent and alternate of ternary expression  multiline-ternary

1:16  error    'd' is not defined                                                       no-undef

1:20  error    'e' is not defined                                                       no-undef

1:22  warning  Expected linebreaks to be 'LF' but found 'CRLF'                          linebreak-style

</code>
</details>

**Comment by Aladdin-ADD** [comment]
Thanks for reporting! I can reproduce it.

**Comment by platinumazure** [comment]
I'm not sure this is a bug. Logical AND has a higher precedence than the conditional operator, so the rule is correctly flagging parentheses around the logical AND as unnecessary.

**Comment by Aladdin-ADD** [comment]
sorry, didn't notice the precedence.  this is working as intended. 😄

**Comment by platinumazure** [comment]
@andoras Thanks for the issue! I'm going to close this as the rule is currently working as intended. If you would like to request an enhancement to this rule, please open a new issue and fill out [this template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md) in order for us to understand how the rule should be enhanced. Thanks!

**Comment by andoras** [comment]
I can understand your argumentation, but in the config i used an exception to this rule and it does not work as expected here. I would expect `"nestedBinaryExpressions": false` to disable the rule for exactly this case, even i would expect this to be legal code at all, because the documentation allows such a construct with `(function(){}) ? a() : b();`. Maybe someone can give me some details what improvement is required to allow this construct?
